pkg/controller/bastion/actuator_reconcile.go

```diff
@@ -0,0 +1,327 @@
+	if err != nil {
+		return err
+	}
+
+	opt, err := DetermineOptions(bastion, cluster)
+	if err != nil {
+		return fmt.Errorf("failed to determine options: %w", err)
+	}
+
+	onmetalClient, namespace, err := onmetal.GetOnmetalClientAndNamespaceFromCloudProviderSecret(ctx, a.Client(), cluster.ObjectMeta.Name)
+	if err != nil {
+		return fmt.Errorf("failed to get onmetal client and namespace from cloudprovider secret: %w", err)
+	}
+	infraStatus, err := getInfrastructureStatus(ctx, onmetalClient, cluster)
+	if err != nil {
+		return fmt.Errorf("failed to get infrastructure status: %w", err)
+	}
+
+	// TODO: Add NetworkPolicy related implementation
+
+	ignitionSecret, err := createIgnitionSecret(namespace, opt)
+	if err != nil {
+		return fmt.Errorf("failed to create ignition secret: %w", err)
+	}
+
+	bastionMachine := createMachine(namespace, a.bastionConfig, infraStatus, opt.BastionInstanceName, ignitionSecret.Name)
+
+	machine, err := applyMachine(ctx, onmetalClient, ignitionSecret, bastionMachine)
+	if err != nil {
+		return fmt.Errorf("failed to create machine: %w", err)
+	}
+
+	endpoints, err := getMachineEndpoints(machine)
+	if err != nil {
+		return fmt.Errorf("failed to get machine endpoints: %w", err)
+	}
+
+	if !endpoints.Ready() {
+		return &reconcilerutils.RequeueAfterError{
+			// requeue rather soon, so that the user (most likely gardenctl eventually)
+			// doesn't have to wait too long for the public endpoint to become available
+			RequeueAfter: 5 * time.Second,
+			Cause:        fmt.Errorf("bastion instance has no public/private endpoints yet"),
+		}
+	}
+
+	// once a public endpoint is available, publish the endpoint on the
+	// Bastion resource to notify upstream about the ready instance
+	log.V(2).Info("Successfully reconciled bastion host")
+	patch := client.MergeFrom(bastion.DeepCopy())
+	bastion.Status.Ingress = endpoints.public
+	return a.Client().Status().Patch(ctx, bastion, patch)
+}
+
+// getMachineEndpoints function returns the bastion endpoints of a running
+// machine. It first validates that the machine is in running state, then
+// extracts the private and public IP of the machine's network interface, and
+// finally converts the IPs to their respective ingress addresses.
+func getMachineEndpoints(machine *computev1alpha1.Machine) (*bastionEndpoints, error) {
+	if machine == nil {
+		return nil, fmt.Errorf("machine can't be nil")
+	}
+
+	if machine.Status.State != computev1alpha1.MachineStateRunning {
+		return nil, fmt.Errorf("machine not running, status: %s", machine.Status.State)
+	}
+
+	endpoints := &bastionEndpoints{}
+
+	if len(machine.Status.NetworkInterfaces) == 0 {
+		return nil, fmt.Errorf("no network interface found for machine: %s", machine.Name)
+	}
+
+	if len(machine.Status.NetworkInterfaces[0].IPs) == 0 {
+		return nil, fmt.Errorf("no private ip found for network interface for machine: %s", machine.Name)
+	}
+
+	privateIP := machine.Status.NetworkInterfaces[0].IPs[0].String()
```

Comment:
We need to make sure that we return here the IPv4 address of the network interface as we can have multiple addresses here. Also please add a TODO here to handle the IPv6 case once the Gardener extension has proper support for IPv6.

Reply:
Done.